pkg/apis/flowcontrol/types.go

```diff
@@ -0,0 +1,298 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package flowcontrol
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+const (
+	FlowDistinguisherSourceUserName          FlowDistinguisherSource = "UserName"
+	FlowDistinguisherSourceUserGroup         FlowDistinguisherSource = "UserGroup"
+	FlowDistinguisherSourceAPIScopeNamespace FlowDistinguisherSource = "Namespace"
+)
+
+const (
+	UserAll  = "*"
+	GroupAll = "*"
+
+	APIGroupAll   = "*"
+	APIVersionAll = "*"
+	ResourceAll   = "*"
+)
+
+// These are valid condition statuses. "ConditionTrue" means a resource is in the condition;
+// "ConditionFalse" means a resource is not in the condition; "ConditionUnknown" means kubernetes
+// can't decide if a resource is in the condition or not. In the future, we could add other
+// intermediate conditions, e.g. ConditionDegraded.
+const (
+	ConditionTrue    ConditionStatus = "True"
+	ConditionFalse   ConditionStatus = "False"
+	ConditionUnknown ConditionStatus = "Unknown"
+)
+
+// +genclient
+// +genclient:nonNamespaced
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// FlowSchema defines the schema of a group of flows. Note that a flow makes up of a set of inbound API requests with
+// similar attributes and is identified by a pair of strings: the name of the FlowSchema and a "flow distinguisher".
+type FlowSchema struct {
+	metav1.TypeMeta
+	// Standard object's metadata.
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#metadata
+	// +optional
+	metav1.ObjectMeta
+	// Specification of the desired behavior of a flow-schema.
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#spec-and-status
+	// +optional
+	Spec FlowSchemaSpec
+	// Current status of a flow-schema.
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#spec-and-status
+	// +optional
+	Status FlowSchemaStatus
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// FlowSchemaList is a list of FlowSchema objects.
+type FlowSchemaList struct {
+	metav1.TypeMeta
+	// Standard list metadata.
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#metadata
+	// +optional
+	metav1.ListMeta
+
+	// Items is a list of flow-schemas.
+	Items []FlowSchema
+}
+
+// FlowSchemaSpec describes how the flow-schema's specification looks like.
+type FlowSchemaSpec struct {
+	// MatchingPriority controls the ordering the flow-schemas in the cluster. Its value must be greater than zero and
+	// also unique. The primary sort is based on MatchingPriority, ordered highest to lowest (20 before 10).
+	MatchingPriority int32
```

Comment:
Must this refer to an existing priority level?

Reply:
no, it doesn't have to. the flow schema referencing a priority level doesn't exist will be ignored (furthermore, we can mark an "unavailable" condition in its status) until the the priority level is re-created.

Reply:
merged replacement